OPNET-582,OPNET-585: Add monitor test for on-prem keepalived

cybertron opened the pull request:
This will provide a timeline of VIP moves while also verifying that the correct priority is present in the keepalived logs.

cybertron commented:
/hold

Looks like I'm still missing a piece to construct the timeline.

cybertron commented:
/test e2e-metal-ipi

cybertron commented:
/test e2e-metal-ipi-serial

cybertron commented:
/test e2e-metal-ipi-ovn-dualstack

cybertron commented:
/test e2e-metal-ipi-ovn-dualstack

cybertron commented:
/hold cancel
/cc @mkowalski 

Apparently the VSphere job in this repo is UPI so there was nothing to be tested, but it looks like it worked fine in OpenStack too so I think we're good to go.

mkowalski commented:
/lgtm

cybertron commented:
/retest-required

cybertron commented:
/retest-required

cybertron commented:
/retest-required

cybertron commented:
/assign @dgoodwin

dgoodwin commented on pkg/monitortests/network/onpremkeepalived/keepalived_priority_check.go:
```diff
@@ -0,0 +1,265 @@
+
+func (w *operatorLogAnalyzer) CollectData(ctx context.Context, storageDir string, beginning, end time.Time) (monitorapi.Intervals, []*junitapi.JUnitTestCase, error) {
+	localRecorder := monitor.NewRecorder()
+	if err := scanAllOperatorPods(ctx, w.kubeClient, newOperatorLogHandlerAfterTime(localRecorder, beginning)); err != nil {
```
Can you add a comment on why you use scanAllOperatorPods both in startcollection and collectdata and how that will work without generating dupes. I don't see dupes, I'm curious why not.

dgoodwin commented on pkg/monitortests/network/onpremkeepalived/keepalived_priority_check.go:
```diff
@@ -0,0 +1,265 @@
+	return localRecorder.Intervals(time.Time{}, time.Time{}), nil, nil
+}
+
+func (*operatorLogAnalyzer) ConstructComputedIntervals(ctx context.Context, startingIntervals monitorapi.Intervals, recordedResources monitorapi.ResourcesMap, beginning, end time.Time) (monitorapi.Intervals, error) {
```
Do you need to make use of beginning and end to prevent picking up events outside the current execution? In an upgrade job openshift-tests is invoked twice, once for the upgrade and once for conformance after. You need to bound the creation of intervals to only those which fit within this execution otherwise the conformance graphs will blow out to include all of upgrade. I didn't see any on-oprem lb upgrade jobs run but you could kick one off with /payload-job.

If it's actually in here and I've missed it then my apologies.

dgoodwin commented on pkg/monitortests/network/onpremkeepalived/keepalived_priority_check.go:
```diff
@@ -0,0 +1,265 @@
+	// Force the test to flake even if it failed
+	ret = append(ret, &junitapi.JUnitTestCase{
+		Name: testName,
+	})
```
From the new test detection: 

"[Jira:"Networking / On-Prem Load Balancer"] on-prem loadbalancer must achieve full priority" [Total: 36, Pass: 36, Fail: 0, Flake: 30]

Anything that can be done about that flake rate? Is that expected at this point? Actually now that I think of it, does this monitortest need to disable itself if it's not on an on-prem platform? I don't see anything in here preventing this from running on the normal clouds.

cybertron reviewed (commented):
I moved this back to draft so I can test the behavior on upgrade jobs before re-running everything.

cybertron commented:
/payload-job periodic-ci-openshift-release-master-nightly-4.18-upgrade-from-stable-4.17-e2e-metal-ipi-upgrade-ovn-ipv6

cybertron commented:
/payload-job periodic-ci-openshift-release-master-nightly-4.18-upgrade-from-stable-4.17-e2e-metal-ipi-upgrade-ovn-ipv6

Previous job failed for what appear to be unrelated reasons and there seem to be no intervals generated.

cybertron commented:
/payload-job periodic-ci-openshift-release-master-nightly-4.19-upgrade-from-stable-4.18-e2e-metal-ipi-upgrade-ovn-ipv6

Same result, trying a different release.

cybertron commented:
/payload-job periodic-ci-openshift-release-master-nightly-4.19-upgrade-from-stable-4.18-e2e-metal-ipi-upgrade-ovn-ipv6

Regular dual stack job passed so I don't think I broke something with the new tests. One more try and then I'll go looking for help on why these upgrade jobs keep failing.

cybertron commented:
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-metal-ipi-upgrade-ovn-ipv6

cybertron commented:
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-metal-ipi-ovn-bm-upgrade

Trying an upgrade job without ipv6 because I'm seeing logs that suggest there might be an issue with it in the payload jobs.

cybertron commented:
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-metal-ipi-ovn-bm-upgrade

cybertron commented:
Okay, upgrade timelines look good now. Our intervals are within the range of the others so we're not going to blow out the display. Still need to verify that it isn't flaking inappropriately, but I'm going to mark this ready for review so we run all the other jobs.

cybertron commented:
No more flaking on cloud jobs and the intervals look good in the jobs I've checked, so I think this is good to go.

mkowalski commented:
/lgtm

cybertron commented:
Okay, I had to make a couple of updates to this due to outside changes. First, I added the PrepareCollection function that is now needed and used it for creating our client. Second, the API priority has changed from 65 to 68 so that test was failing on recent builds.

I also added an ingress VIP test because we ran into a bug with that as well. It works exactly the same as the API test, just looking for the ingress priority instead.

dgoodwin commented:
/lgtm

cybertron commented:
/hold cancel

The previous issue preventing this from merging has now been fixed.
